src/Function.cs

```diff
@@ -682,94 +500,88 @@ internal Function(Interop.StoreHandle store, Delegate callback, bool hasReturn)
                 parameterTypes = parameterTypes[1..];
             }
 
-            Parameters = GetParameterKinds(parameterTypes);
-            Results = GetReturnTypeKinds(EnumerateReturnTypes(returnType));
+            AddParameters(parameterTypes);
+            AddResults(EnumerateReturnTypes(returnType));
 
-            var parameters = CreateValueTypeVec(Parameters);
-            var results = CreateValueTypeVec(Results);
-            using var funcType = Interop.wasm_functype_new(ref parameters, ref results);
+            var parameters = new ValueTypeArray(Parameters);
+            var results = new ValueTypeArray(Results);
+            using var funcType = new TypeHandle(Native.wasm_functype_new(parameters, results));
 
-            if (hasCaller)
-            {
-                var func = CreateWasmtimeCallback(store, callback, parameterTypes.Length, Results);
-                Handle = Interop.wasmtime_func_new_with_env(
-                    store,
-                    funcType,
-                    func,
-                    GCHandle.ToIntPtr(GCHandle.Alloc(func)),
-                    Interop.GCHandleFinalizer
-                );
-            }
-            else
+            unsafe
             {
-                var func = CreateCallback(store, callback, parameterTypes.Length, Results);
-                Handle = Interop.wasm_func_new_with_env(
-                    store,
+                Native.WasmtimeFuncCallback? func = null;
+                if (hasCaller)
+                {
+                    func = (env, caller, args, nargs, results, nresults) =>
+                        InvokeCallback(callback, new Caller(caller), true, args, (int)nargs, results, (int)nresults, Results);
+                }
+                else
+                {
+                    func = (env, caller, args, nargs, results, nresults) =>
+                        InvokeCallback(callback, new Caller(caller), false, args, (int)nargs, results, (int)nresults, Results);
```

Comment:
In other bindings I have a `finally` block which nulls out the interior pointer in `Caller` to signify that it's no longer valid to use (ideally preventing accidental bugs where the `Caller` object is stored to outlive the call and then it's used again). 

(e.g. something [like this](https://github.com/bytecodealliance/wasmtime-py/blob/64f5145a0ed5a5d8afe2203f46a25ae9303c9713/wasmtime/_func.py#L215-L216))

Reply:
Changing `Caller` to `ref struct` should prevent storage of it beyond this stack frame.  I can make that fix.

Reply:
Although I just remember why I can't do that as we invoke the .NET callback via reflection currently, so that can't happen if it's `ref struct`.

I'll go with invalidating the interior pointer and throwing an exception when accessed again.